File: services/grid-service/pkg/mdf/v1/molecule.go
package mdfv1

import (
	"time"
)

// Molecule represents the atomic unit of data in QuantatomAI (MDF).
// This struct is designed to map cleanly to Parquet columns.
type Molecule struct {
	// 1. The Coordinates (The "Where")
	CoordinateHash   []byte            `parquet:"coordinate_hash"`   // 128-bit MurmurHash3
	CustomDimensions map[string]string `parquet:"custom_dimensions"` // Map for sparse dims

	// 2. The Payload (The "What") - Polymorphic
	// Parquet handles these as separate columns, usually null if not set.
	NumericValue    *float64 `parquet:"numeric_value,optional"`
	TextCommentary  *string  `parquet:"text_commentary,optional"`
	EmbeddingVector []byte   `parquet:"embedding_vector,optional"` // Blob for simplicity in Parquet

	// Ultra Diamond Upgrade: Rich Types
	DateValue    *int64  `parquet:"date_value,optional"` // Unix Millis
	BooleanValue *bool   `parquet:"boolean_value,optional"`
	ErrorValue   *string `parquet:"error_value,optional"`

	// 3. The Context (The "Why")
	Timestamp      int64  `parquet:"timestamp"`
	SourceSystem   string `parquet:"source_system"`
	SecurityMask   uint64 `parquet:"security_mask"`
	CausalityClock []byte `parquet:"causality_clock"`
	IsLocked       bool   `parquet:"is_locked"` // Prevent Top-Down Overwrites
}

// NewNumericMolecule creates a standard numeric data point.
func NewNumericMolecule(hash []byte, dims map[string]string, val float64, source string, mask uint64) *Molecule {
	v := val
	return &Molecule{
		CoordinateHash:   hash,
		CustomDimensions: dims,
		NumericValue:     &v,
		Timestamp:        time.Now().UnixNano(),
		SourceSystem:     source,
		SecurityMask:     mask,
	}
}

// NewCommentaryMolecule creates a text annotation.
func NewCommentaryMolecule(hash []byte, dims map[string]string, comment string, source string, mask uint64) *Molecule {
	c := comment
	return &Molecule{
		CoordinateHash:   hash,
		CustomDimensions: dims,
		TextCommentary:   &c,
		Timestamp:        time.Now().UnixNano(),
		SourceSystem:     source,
		SecurityMask:     mask,
	}
}

File: services/atom-engine/src/mdf/molecule.rs

use arrow::datatypes::{DataType, Field, Schema};
use std::sync::Arc;

pub struct MoleculeSchema;

impl MoleculeSchema {
    pub fn schema() -> Arc<Schema> {
        Arc::new(Schema::new(vec![
            Field::new("coordinate_hash", DataType::Binary, false),
            // Map is complex in Arrow, often represented as List of Structs. 
            // For simplicity in v1, we might treat custom_dimensions as List<Struct<Key, Value>>
            // But Parquet Go writer writes Map. Arrow reader should handle it.
            // Let's rely on schema inference from file for now.
            // Field::new("custom_dimensions", DataType::Map(...), true), 
            
            Field::new("numeric_value", DataType::Float64, true),
            Field::new("text_commentary", DataType::Utf8, true),
            Field::new("embedding_vector", DataType::Binary, true),
            
            // Ultra Diamond Upgrade: Rich Types
            Field::new("date_value", DataType::Int64, true), // Unix Millis
            Field::new("boolean_value", DataType::Boolean, true),
            Field::new("error_value", DataType::Utf8, true),
            
            Field::new("timestamp", DataType::Int64, false),
            Field::new("source_system", DataType::Utf8, false),
            Field::new("security_mask", DataType::UInt64, false),
            Field::new("causality_clock", DataType::Binary, true),
            
            // Collision Resolution (Red Team Audit Response)
            Field::new("is_locked", DataType::Boolean, true),
        ]))
    }
}
